x-pack/plugins/apm/server/lib/apm_client/index.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { CoreSetup } from 'src/core/server';
+import { ESSearchBody } from '../../../typings/elasticsearch';
+import { ApmIndicesConfig } from '../settings/apm_indices/get_apm_indices';
+
+// TODO: should we use ApmIndicesConfig here instead of these custom names?
+type Index = 'transaction' | 'span' | 'error' | 'metrics';
+
+interface Search {
+  apmIndices: Index[];
+  body: ESSearchBody;
+}
+
+interface ApmClient {
+  search: (params: Search) => void;
+}
+
+const indexMap: Record<Index, keyof ApmIndicesConfig> = {
+  transaction: 'apm_oss.transactionIndices',
+  error: 'apm_oss.errorIndices',
+  span: 'apm_oss.spanIndices',
+  metrics: 'apm_oss.metricsIndices',
+};
+
+export let apmClient: ApmClient = {
+  search() {
+    throw new Error("apmClient wasn't initialized");
+  },
+};
+export function createApmClient(core: CoreSetup) {
+  apmClient = {
+    search: async ({ apmIndices, body }: Search) => {
+      const indices = apmIndices.map((index: Index) => indexMap[index]);
+      // TODO: callAsCurrentUser isn't available here. How to get it?
+      const { callAsInternalUser } = core.elasticsearch.legacy.client;
```

Comment:
I think you need to use `RequestHandlerContext` like we do in https://github.com/elastic/kibana/blob/fa8187ba2a1779e1f74c215a6862b0a4825d226b/x-pack/plugins/apm/server/lib/helpers/es_client.ts#L87-L116.

I think for querying public apm indices (not system indices) the current user should be used. For querying internal indices like `.apm-agent-configuration` or `.apm-custom-links` the internal user should be used.
Thinking more about this, it should probably be separate clients.


So the following client should always use the current user, and should not be used for querying system indices:
```js
apmClient.search({
  apmIndices: ['transaction', 'span'],
  size: 100,
  query: {...},
  aggs: {}
})
```

The following client should always use the internal user, but only be used to querying system indices
```js
apmInternalClient.search({
  apmIndices: ['.apm-agent-configuration', '.apm-custom-links'],
  size: 100,
  query: {...},
  aggs: {}
})
```

Reply:
The "normal" `apmClient` should filter out legacy data `addFilterForLegacyData` and handle frozen indices. `apmInternalClient` should not.

Reply:
Currently we have one client that all requests are made with. I think we should go away from that thinking and instead have multiple smaller clients, eg `apmClient`, `apmInternalClient`, `mlClient` (until they ship their own) etc.

Reply:
It would be great if you can update `es_client` to reflect this change.

Reply:
I think you can `apmClient` a lot since it should only have a `search` method (no `delete`, `index` etc).